source/dynamical_systems/src/Ring.cpp

```diff
@@ -0,0 +1,240 @@
+
+  if (pose.get_orientation().dot(qd) < 0) {
+    qd.coeffs() = -qd.coeffs();
+  }
+
+  Eigen::Quaterniond deltaQ = qd * pose.get_orientation().conjugate();
+  if (deltaQ.vec().norm() < 1e-7) {
+    return localAngularVelocity;
+  }
+
+  //dOmega = 2 * ln (deltaQ)
+  Eigen::Quaterniond deltaOmega = Eigen::Quaterniond::Identity();
+  deltaOmega.w() = 0;
+  double phi = atan2(deltaQ.vec().norm(), deltaQ.w());
+  deltaOmega.vec() = 2 * deltaQ.vec() * phi / sin(phi);
+
+  localAngularVelocity = get_angular_gain() * deltaOmega.vec();
+  localAngularVelocity *= localFieldStrength;
+
+  Eigen::Vector2d position2d(pose.get_position().x(), pose.get_position().y());
+  Eigen::Vector2d linearVelocity2d(linearVelocity.x(), linearVelocity.y());
+  if (position2d.norm() < 1e-7 || linearVelocity2d.norm() < 1e-7) {
+    return localAngularVelocity;
+  }
+
+  double projection = position2d.normalized().dot((position2d + linearVelocity2d).normalized());
+  double dThetaZ = 0;
+  if (1 - abs(projection) > 1e-7) {
+    dThetaZ = acos(projection);
+  }
+  localAngularVelocity.z() += dThetaZ;
+
+  return localAngularVelocity;
+}
+
+void Ring::set_base_frame(const state_representation::CartesianState& base_frame) {
+  DynamicalSystem<state_representation::CartesianState>::set_base_frame(base_frame);
+  // update reference frame of center
+  auto center = this->get_center();
+  center.set_reference_frame(base_frame.get_name());
+  this->set_center(center);
+  // update reference frame of rotation offset
+  this->set_rotation_offset(this->get_rotation_offset());
+}
+
+void Ring::set_center(const CartesianPose& center) {
+  // validate that the reference frame of the center is always compatible with the DS reference frame
+  if (center.get_reference_frame() != this->get_base_frame().get_name()) {
+    if (center.get_reference_frame() != this->get_base_frame().get_reference_frame()) {
+      throw state_representation::exceptions::IncompatibleReferenceFramesException(
+          "The reference frame of the center " + center.get_name() + " in frame " + center.get_reference_frame()
+              + " is incompatible with the base frame of the dynamical system "
+              + this->get_base_frame().get_name() + " in frame " + this->get_base_frame().get_reference_frame() + "."
+      );
+    }
+    this->center_->set_value(this->get_base_frame().inverse() * center);
+  } else {
+    this->center_->set_value(center);
+  }
+}
+
+void Ring::set_rotation_offset(const Eigen::Quaterniond& rotation) {
+  auto pose = CartesianPose::Identity("rotation", this->get_base_frame().get_name());
+  pose.set_orientation(rotation);
+  this->rotation_offset_->set_value(pose);
+}
+
+void Ring::set_radius(double radius) {
+  this->radius_->set_value(radius);
+}
+
+void Ring::set_width(double width) {
+  this->width_->set_value(width);
+}
+
+void Ring::set_speed(double speed) {
+  this->speed_->set_value(speed);
+}
+
+void Ring::set_field_strength(double field_strength) {
```

Comment:
I believe I've mentioned before that I'm not a huge fan of excessive inlining and function definitions in headers. Compilers gonna do what compilers gonna do, so I prefer to make the public library headers smaller

Reply:
I totally see that. For me it was just not sure that the compiler is actually inlining functions that are defined in a separate source file as we have. The thread I have linked is not giving a clear answer on that question. If it does then yes we should remove all the inlines it would be much cleaner and not much of refactoring to do actually. Then the only acceptable case where we have functions defined in the header would be template as this is the proper way.